includes/class-wc-payments-webhook-processing-service.php

```diff
@@ -366,18 +366,34 @@ private function process_webhook_payment_intent_succeeded( $event_body ) {
 		$event_data    = $this->read_webhook_property( $event_body, 'data' );
 		$event_object  = $this->read_webhook_property( $event_data, 'object' );
 		$intent_id     = $this->read_webhook_property( $event_object, 'id' );
+		$currency      = $this->read_webhook_property( $event_object, 'currency' );
 		$order         = $this->get_order_from_event_body_intent_id( $event_body );
 		$intent_status = $this->read_webhook_property( $event_object, 'status' );
 		$event_charges = $this->read_webhook_property( $event_object, 'charges' );
 		$charges_data  = $this->read_webhook_property( $event_charges, 'data' );
 		$charge_id     = $this->read_webhook_property( $charges_data[0], 'id' );
 
+		$payment_method_id = $charges_data[0]['payment_method'] ?? null;
 		if ( ! $order ) {
 			return;
 		}
-		// update _charge_id meta if it doesn't exist - happens when maybe_process_upe_redirect fails sometimes.
-		if ( $charge_id && ! $order->get_meta( '_charge_id' ) ) {
-			$order->update_meta_data( '_charge_id', $charge_id );
+		// Update missing intents because webhook can be delivered before order is processed on the client.
+		$meta_data_to_update = [
+			'_intent_id'         => $intent_id,
+			'_charge_id'         => $charge_id,
+			'_payment_method_id' => $payment_method_id,
+			WC_Payments_Utils::ORDER_INTENT_CURRENCY_META_KEY => $currency,
+		];
+
+		$order_changed = false;
+		foreach ( $meta_data_to_update as $key => $value ) {
+			if ( $value && ! $order->get_meta( $key ) ) {
```

Comment:
Hey @zmaglica - Sorry for chiming in a bit late, but what's the reasoning behind skipping the update if the order has something in the meta?

I ask because of p1654285164749429/1654031318.105099-slack-CGGCLBN58, in which orders were left with a stale `_charge_id` instead of the successful one, because of the `! $order->get_meta( '_charge_id' )` check. Is checking if they match, and updating the meta if they don't, a feasible approach here?

Reply:
I wasn't sure where what is the correct source of data, so I skipped updating if the value existed. What I did here is the same this as it was before, but I added checks for all meta keys that can be found in intent.

For example:

Before this, the code that was updating the charge_id was:

```
if ( $charge_id && ! $order->get_meta( '_charge_id' ) ) {
			$order->update_meta_data( '_charge_id', $charge_id );
}
```
As you can see, the code is the same, since `charge_id` is the part of meta keys that needs to be updated. I can easily remove `$order->get_meta( $key )` if you think that value that comes from webhook is always correct.

Reply:
> As you can see, the code is the same, since charge_id is the part of meta keys that needs to be updated.

Yep, I had noticed that. The previous `! $order->get_meta( '_charge_id' )` check is what caused the old charge ID to not be updated in the case I investigated in that Slack thread.

> I can easily remove $order->get_meta( $key ) if you think that value that comes from webhook is always correct.

I've taken a look at the code, and when processing the order we always seem to override the meta, like [here](https://github.com/Automattic/woocommerce-payments/blob/114f6126ace6968837a0045315a6561ed3fc67ee/includes/payment-methods/class-upe-payment-gateway.php#L646) when called from `maybe_process_upe_redirect` for example.

I think it's likely that Stripe will have the most up-to-date data, especially considering that most UPE processing happens on their end. However, I'm not sure if we make some assumption somewhere in the plugin that could break if we start updating the data when Stripe does, so perhaps this needs more investigating or even some architecture's guild input. I'll create a new issue to track this and ask for some more feedback before we remove the check.

Reply:
We can update charge_id for every webhook request if needed, and leave the reaming ones to be updated if missing. Will that be a viable option for this use case?